matchers/parse/parse.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		if errors.Is(err, ErrEOF) {
+			// If this is the end of input we still need to check if the optional
+			// open brace has a matching close brace
+			return p.parseCloseBrace, nil
+		}
+		return nil, fmt.Errorf("%s: %s", err, "expected a matcher or close brace after comma")
+	}
+	if tok.Kind == TokenCloseBrace {
+		return p.parseCloseBrace, nil
+	}
+	return p.parseMatcher, nil
+}
+
+func (p *Parser) parseEOF(l *Lexer) (parseFunc, error) {
+	tok, err := l.Scan()
+	if err != nil {
+		return nil, fmt.Errorf("%s: %w", err, ErrExpectedEOF)
+	}
+	if !tok.IsEOF() {
+		return nil, fmt.Errorf("%d:%d: %s: %w", tok.ColumnStart, tok.ColumnEnd, tok.Value, ErrExpectedEOF)
+	}
+	return nil, nil
+}
+
+// accept returns true if the next token is one of the specified kinds,
+// otherwise false. If the token is accepted it is consumed. TokenEOF is
+// not an accepted kind  and instead accept returns ErrEOF if there is no
+// more input.
+func (p *Parser) accept(l *Lexer, kinds ...TokenKind) (ok bool, err error) {
+	ok, err = p.acceptPeek(l, kinds...)
+	if ok {
+		if _, err = l.Scan(); err != nil {
+			panic("Failed to scan peeked token, this should never happen")
+		}
+	}
+	return ok, err
+}
+
+// acceptPeek returns true if the next token is one of the specified kinds,
+// otherwise false. However, unlike accept, acceptPeek does not consume accepted
+// tokens. TokenEOF is not an accepted kind and instead accept returns ErrEOF
+// if there is no more input.
+func (p *Parser) acceptPeek(l *Lexer, kinds ...TokenKind) (bool, error) {
+	tok, err := l.Peek()
+	if err != nil {
+		return false, err
+	}
+	if tok.IsEOF() {
+		return false, ErrUnexpectedEOF(l)
+	}
+	return tok.IsOneOf(kinds...), nil
+}
+
+// expect returns the next token if it is one of the specified kinds, otherwise
+// it returns an error. If the token is expected it is consumed. TokenEOF is not
+// an accepted kind and instead expect returns ErrEOF if there is no more input.
+func (p *Parser) expect(l *Lexer, kind ...TokenKind) (Token, error) {
+	tok, err := p.expectPeek(l, kind...)
+	if err != nil {
+		return tok, err
+	}
+	if _, err = l.Scan(); err != nil {
+		panic("Failed to scan peeked token, this should never happen")
+	}
+	return tok, nil
+}
+
+// expect returns the next token if it is one of the specified kinds, otherwise
+// it returns an error. However, unlike expect, expectPeek does not consume tokens.
+// TokenEOF is not an accepted kind and instead expect returns ErrEOF if there is no
+// more input.
+func (p *Parser) expectPeek(l *Lexer, kind ...TokenKind) (Token, error) {
+	tok, err := l.Peek()
+	if err != nil {
+		return tok, err
+	}
+	if tok.IsEOF() {
+		return tok, ErrUnexpectedEOF(l)
+	}
+	if !tok.IsOneOf(kind...) {
+		return tok, fmt.Errorf("%d:%d: unexpected %s", tok.ColumnStart, tok.ColumnEnd, tok.Value)
+	}
+	return tok, nil
+}
+
+// Unquote the value in token. If unquoted returns it unmodified.
+func (p *Parser) unquote(t Token) (string, error) {
```

Comment:
This does not uses `p` can it be a regular function instead?

Reply:
On further thought I think your original suggestion where `Unquote` is a method of Token is best, I'm going to use that instead.